packages/nextra/src/loader.ts

```diff
@@ -208,39 +231,55 @@ export default MDXContent`
       // Remove the only `index` route
       .replace(/^index$/, '')
 
+  const stringifiedPageOpts = JSON.stringify(pageOpts)
+  const pageOptsChecksum = hashFnv32a(stringifiedPageOpts)
+
   return `import { SSGContext as __nextra_SSGContext__ } from 'nextra/ssg'
 ${themeConfigImport}
 ${cssImport}
 
-const __nextra_pageOpts__ = ${JSON.stringify(pageOpts)}
+const __nextra_pageOpts__ = ${stringifiedPageOpts}
 
-globalThis.__nextra_internal__ = {
-  pageMap: __nextra_pageOpts__.pageMap,
-  route: __nextra_pageOpts__.route
-}
+// Make sure the same component is always returned so Next.js will render the
+// stable layout. We then put the actual content into a global store and use
+// the route to identify it.
+const NEXTRA_INTERNAL = Symbol.for('__nextra_internal__')
+const __nextra_internal__ = (globalThis[NEXTRA_INTERNAL] ||= Object.create(null))
+__nextra_internal__.pageMap = __nextra_pageOpts__.pageMap
+__nextra_internal__.route = __nextra_pageOpts__.route
+__nextra_internal__.context ||= Object.create(null)
+__nextra_internal__.refreshListeners ||= Object.create(null)
 
 ${result}
 
-__nextra_pageOpts__.title =
-  ${JSON.stringify(frontMatter.title)} ||
-  (typeof __nextra_title__ === 'string' && __nextra_title__) ||
-  ${JSON.stringify(title /* Fallback as sidebar link name */)}
-
 const Content = props => (
   <__nextra_SSGContext__.Provider value={props}>
     <MDXContent />
   </__nextra_SSGContext__.Provider>
 )
 
-globalThis.__nextra_pageContext__ ||= Object.create(null)
+__nextra_pageOpts__.title =
+  ${JSON.stringify(frontMatter.title)} ||
+  (typeof __nextra_title__ === 'string' && __nextra_title__) ||
+  ${JSON.stringify(title /* Fallback as sidebar link name */)}
 
-// Make sure the same component is always returned so Next.js will render the
-// stable layout. We then put the actual content into a global store and use
-// the route to identify it.
-globalThis.__nextra_pageContext__[${JSON.stringify(pageNextRoute)}] = {
+__nextra_internal__.context[${JSON.stringify(pageNextRoute)}] = {
   Content,
   pageOpts: __nextra_pageOpts__,
-  themeConfig: ${themeConfigImport ? '__nextra_themeConfig__' : 'null'}
+  themeConfig: ${themeConfigImport ? '__nextra_themeConfig__' : 'null'},
+}
+${'' /* TODO: Do not include all the HMR logic for prod */}
+if (module.hot) {
```

Comment:
To remove todo comment?
```suggestion
if (process.env.NODE_ENV !== 'production' && module.hot) {
```